### PR TITLE
Respect `PURE_PYTHON` environment variable set to `0`

### PR DESCRIPTION
**Pull request by perrinjerome**
Setting to 0 should be equivalent to not setting the variable.

Refs https://github.com/zopefoundation/meta/issues/283

### CODE REVIEW

**Comment by perrinjerome** [comment]
oh tests are failing, I set to draft

**Review by d-maurer — CHANGES_REQUESTED** [review]
The checker module contains a docstring specifying that the value of the `PURE_PYTHON` envvar is not relevant. It should get changed to be in line with this modification.

**Comment by perrinjerome** [comment]
> The checker module contains a docstring specifying that the value of the `PURE_PYTHON` envvar is not relevant. It should get changed to be in line with this modification.

Thank you, I have changed this.

> oh tests are failing, I set to draft

actually it was OK, I opened the same pull request for another package, but tests here seems to be OK

File: src/zope/security/checker.py
Comment (d-maurer) [inline src/zope/security/checker.py:27]: You now expect the value to be an integer.

I suggest to allow (for all affected packages): interpret a missing, empty an "0" value for the envvar `PURE_PYTHON` as "not PURE_PYTHON". The logic could look something like `PURE_PYTHON = int(os.environ.get("PURE_PYTHON") or 0)`

**Review by d-maurer — CHANGES_REQUESTED** [review]

**Comment by perrinjerome** [comment]
> 
actually it was OK, I opened the same pull request for another package, but tests here seems to be OK

 ... sorry . No, the tests also fail here. It seems there is a circular dependency between https://github.com/zopefoundation/zope.proxy/pull/71 and this one, or something I don't understand yet.

**Comment by d-maurer** [comment]
Jérome Perrin wrote at 2024-9-30 00:04 -0700:
> ... sorry . No, the tests also fail here. It seems there is a circular dependency between https://github.com/zopefoundation/zope.proxy/pull/71 and this one, or something I don't understand yet.

I think the test is wrong:
it import `_Proxy` (which is the C implementation) but apparently,
it is not available.
Instead, the test should import "Proxy", which is
the C implementation if available otherwise the Python implementation.

**Comment by d-maurer** [comment]
> Instead, the test should import "Proxy", which is the C implementation if available otherwise the Python implementation.

If we do this, we may want to add an additional test: if `not PURE_PYTHON`, then we should test that the C implementation is available. Currently, `zope.security` automatically switches to the Python implementation when the C implementation is unavailable. The test should ensure that this is only the case when we expect it to be so.

File: src/zope/security/checker.py
Comment (perrinjerome) [reply src/zope/security/checker.py]: I pushed a new change to change the wording to make it more clear that an int is expected.

> If the environment variable ``PURE_PYTHON`` is set to ``1`` before this module is imported

In my understanding, all affected packages use a logic similar to `int(os.environ.get("PURE_PYTHON") or 0`

**Comment by perrinjerome** [comment]
I understood the problem with the test, it's because the changes from this pull request depend on https://github.com/zopefoundation/zope.proxy/pull/71 - there's a circular dependency between these two pull requests. The solution seems to be to merge both and release both packages.

Also, the test importing `from zope.security.proxy import _Proxy` seems correct, this is the test for the C version so it uses the full name as the target class, not the alias, to be sure to test the C class. The equivalent python test also uses the full name of the equivalent python class (`from zope.security.proxy import ProxyPy`) and the alias is not used here. This seems correct after all.

---

The long story about the analysis of the [test errors](https://github.com/zopefoundation/zope.security/actions/runs/11100803261/job/30837281563?pr=115#step:12:448) `ImportError: cannot import name '_Proxy' from 'zope.security.proxy' (/io/.tox/py310/lib/python3.10/site-packages/zope/security/proxy.py)`. This error happens here when running with `PURE_PYTHON=0`. In that test run, `zope.security` understands `PURE_PYTHON=0` as "use C", but `zope.proxy` does not understands it and consider as "do not use C", because it is the current version from pypi without the changes from https://github.com/zopefoundation/zope.proxy/pull/71 


The test code is here:

https://github.com/zopefoundation/zope.security/blob/e93ff5f4b202354e3b7be46fbabfd5ec2fcdbca1/src/zope/security/tests/test_proxy.py#L1284-L1291

This test is not skipped because `PURE_PYTHON` is False. 

`from zope.security.proxy import _Proxy` is re-exported here:

https://github.com/zopefoundation/zope.security/blob/66d0c3d72b7066783762218c7e6b4c933d275a30/src/zope/security/proxy.py#L352-L357

The module import happens in C here:

https://github.com/zopefoundation/zope.security/blob/66d0c3d72b7066783762218c7e6b4c933d275a30/src/zope/security/_proxy.c#L860-L861

https://github.com/zopefoundation/zope.security/blob/869deee11838b8617998969d98e65b0f868c8629/include/zope.proxy/zope/proxy/proxy.h#L27-L43

In plain python, this would be `from zope.proxy import _CAPI`.

And in `zope.proxy`, this is still checking `if 'PURE_PYTHON' not in os.environ:` here

https://github.com/zopefoundation/zope.proxy/blob/cbb23c424826d89538b90bb17dc35d48bacc356a/src/zope/proxy/__init__.py#L483

and once `zope.proxy` is updated to use `if not int(os.environ.get('PURE_PYTHON', '0')):` the test will pass again.

**Comment by perrinjerome** [comment]
In short, what we can do is merge https://github.com/zopefoundation/zope.security/pull/115 and https://github.com/zopefoundation/zope.proxy/pull/71 although test fail for both pull requests, once boths are merged and released, the test should pass again.

**Comment by perrinjerome** [comment]
I pushed a tox.ini change to use zope.proxy from https://github.com/zopefoundation/zope.proxy/pull/71 , this change is not planned to be merged, it's just to be sure that tests will be OK with this change.

File: src/zope/security/checker.py
Comment (d-maurer) [inline src/zope/security/checker.py:27]: I am not yet fully happy (the text still does not yet tell that the envvar is expected to have an `int` value) but the text is not important enough to prevent a merge.

File: tox.ini
Comment (d-maurer) [inline tox.ini:21]: Are we sure we want tests done with a fixed version of `zope.proxy`?

I understand that you had a need to make tests against that fixed version to analyze the observed test failures (which you now found to have been caused by the released `zope.proxy` version to interpret `PURE_PYTHON` still in the "old" way), but please do not forget to remove this test version pin again once a new `zope.proxy` is released.

**Review by d-maurer — APPROVED** [review]

File: tox.ini
Comment (perrinjerome) [reply tox.ini:21]: You are right, this is only to run the tests before new `zope.proxy` is released. This pull request is still draft, once the new `zope.proxy` is released I will revert this change and mark it ready for review. 

I was thinking we can proceed with the steps described in https://github.com/zopefoundation/zope.proxy/pull/71#issuecomment-2387483569

File: src/zope/security/checker.py
Comment (perrinjerome) [reply src/zope/security/checker.py:27]: I don't like to leave it like that, but I could not find better words

**Comment by perrinjerome** [comment]
I cannot merge this one either because of the test status, but I believe it is in a state where it can be merged.
/cc @dataflake

**Comment by d-maurer** [comment]
Jérome Perrin wrote at 2024-10-2 05:44 -0700:
>@perrinjerome commented on this pull request.
>
>
>
>> @@ -24,10 +24,10 @@
> (:func:`getCheckerForInstancesOf`, :func:`selectChecker`).
>
> This module is accelerated with a C implementation on CPython by
>-default. If the environment variable ``PURE_PYTHON`` is set (to any
>-value) before this module is imported, the C extensions will be
>-bypassed and the reference Python implementations will be used. This
>-can be helpful for debugging and tracing.
>+default. If the environment variable ``PURE_PYTHON`` is set to ``1``
>
>I don't like to leave it like that, but I could not find better words

...environment variable ``PURE_PYTHON`` (converted to ``int``) has
a nonzero value ...

**Comment by perrinjerome** [comment]
> ...environment variable ``PURE_PYTHON`` (converted to ``int``) has a nonzero value ..

Thank you, I did not notice your message before this was merged. I made a PR with your suggestion https://github.com/zopefoundation/zope.security/pull/116

**Pull request merged**
